File: CIMWeb/src/main/java/ch/ninecode/cim/cimweb/ListFiles.java
package ch.ninecode.cim.cimweb;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.logging.Logger;

import javax.ejb.Stateless;
import javax.json.JsonStructure;
import javax.resource.ResourceException;
import javax.resource.cci.Connection;
import javax.resource.cci.Interaction;
import javax.resource.cci.MappedRecord;
import javax.ws.rs.*;

import ch.ninecode.cim.connector.CIMInteractionSpec;
import ch.ninecode.cim.connector.CIMInteractionSpecImpl;
import ch.ninecode.cim.connector.CIMMappedRecord;

/**
 * List the files in HDFS.
 */
@Stateless
@Path("list")
public class ListFiles extends RESTful
{
    static protected String LOGGER_NAME = ListFiles.class.getName ();
    static protected Logger _Logger = Logger.getLogger (LOGGER_NAME); // , String resourceBundleName)

    @GET
    @Produces ({"application/json"})
    public String listFiles (@DefaultValue("false") @MatrixParam ("debug") String debug)
    {
        return (listFiles ("", debug));
    }

    @SuppressWarnings ("unchecked")
    @GET @Path("{path}")
    @Produces ({"application/json"})
    public String listFiles (@PathParam("path") String path, @DefaultValue("false") @MatrixParam ("debug") String debug)
    {
        RESTfulResult ret = new RESTfulResult ();

        Connection connection = getConnection (ret);
        if (null != connection)
        {
            try
            {
                final CIMInteractionSpecImpl spec = new CIMInteractionSpecImpl ();
                spec.setFunctionName (CIMInteractionSpec.LIST_FILES);
                final MappedRecord input = getConnectionFactory ().getRecordFactory ().createMappedRecord (CIMMappedRecord.INPUT);
                input.setRecordShortDescription ("the directory path to list");
                input.put ("path", path.startsWith ("/") ? path : "/" + path);
                if (!debug.equals ("false"))
                    input.put ("debug", debug);
                final MappedRecord output = getConnectionFactory ().getRecordFactory ().createMappedRecord (CIMMappedRecord.OUTPUT);
                output.setRecordShortDescription ("the results of the list operation");
                final Interaction interaction = connection.createInteraction ();
                if (interaction.execute (spec, input, output))
                {
                    if (!output.isEmpty ())
                    {
                        Object result = output.get ("files");
                        if (result instanceof JsonStructure)
                            ret.setResult ((JsonStructure)result);
                        else
                            ret.setResult (result.toString ());
                    }
                    else
                        ret._Message += "interaction returned empty";
                }
                else
                {
                    _Logger.severe ("interaction execution failed");
                    ret._Status = FAIL;
                    ret._Message += "interaction execution failed";
                }
                interaction.close ();
                connection.close ();
            }
            catch (ResourceException resourceexception)
            {
                _Logger.severe ("interaction resource exception");
                StringWriter string = new StringWriter ();
                string.append ("ResourceException on interaction\n");
                PrintWriter writer = new PrintWriter (string);
                resourceexception.printStackTrace (writer);
                ret._Status = FAIL;
                ret._Message = string.toString ();
                writer.close ();
            }
            finally
            {
                try
                {
                    connection.close ();
                }
                catch (ResourceException resourceexception)
                {
                    _Logger.severe ("close resource exception");
                    StringWriter string = new StringWriter ();
                    string.append ("ResourceException on close\n");
                    PrintWriter writer = new PrintWriter (string);
                    resourceexception.printStackTrace (writer);
                    ret._Status = FAIL;
                    ret._Message = string.toString ();
                    writer.close ();
                }
            }
        }
        else
        {
            _Logger.severe ("_CIMConnectionFactory.getConnection() is null");
            ret._Status = FAIL;
            ret._Message += "getConnection failed";
        }

        return (ret.toString ());
    }
}

File: CIMWeb/src/main/scala/ch/ninecode/cim/cimweb/ListFiles.scala
package ch.ninecode.cim.cimweb

import java.util.logging.Logger
import javax.ejb.Stateless
import javax.json.JsonObject
import javax.json.Json
import javax.json.JsonStructure
import javax.resource.ResourceException
import javax.ws.rs.core.MediaType
import javax.ws.rs.DefaultValue
import javax.ws.rs.GET
import javax.ws.rs.MatrixParam
import javax.ws.rs.Path
import javax.ws.rs.PathParam
import javax.ws.rs.Produces

import scala.collection.JavaConversions._
import ch.ninecode.cim.connector.CIMFunction
import ch.ninecode.cim.connector.CIMInteractionSpec
import ch.ninecode.cim.connector.CIMInteractionSpecImpl
import ch.ninecode.cim.connector.CIMMappedRecord

/**
 * List the files in HDFS.
 */
@Stateless
@Path ("list/")
class ListFiles extends RESTful
{
    import ListFiles._

    @GET
    @Produces (Array (MediaType.APPLICATION_JSON))
    def listFiles (@DefaultValue ("false") @MatrixParam ("debug") debug: String): String =
        listFiles ("", debug)

    @GET
    @Path ("{path:[^;]*}")
    @Produces (Array (MediaType.APPLICATION_JSON))
    def listFiles (
        @PathParam ("path") path: String,
        @DefaultValue ("false") @MatrixParam ("debug") debug: String): String =
    {
        val ret = new RESTfulJSONResult
        val directory = if (path.startsWith ("/")) path else "/" + path
        _Logger.info ("list %s".format (directory))
        val function = ListFilesFunction (directory, try { debug.toBoolean } catch { case _: Throwable => false })
        val connection = getConnection (ret)
        if (null != connection)
            try
            {
                val spec: CIMInteractionSpec = new CIMInteractionSpecImpl
                spec.setFunctionName (CIMInteractionSpec.EXECUTE_CIM_FUNCTION)
                val input = getInputRecord ("input record containing the function to run")
                input.asInstanceOf[map].put (CIMFunction.FUNCTION, function)
                val interaction = connection.createInteraction
                val output = interaction.execute (spec, input)
                if (null == output)
                    throw new ResourceException ("null is not a MappedRecord")
                else
                {
                    val record = output.asInstanceOf [CIMMappedRecord]
                    ret.setResult (record.get (CIMFunction.RESULT).asInstanceOf [JsonStructure])
                    val response = ret.result.asInstanceOf[JsonObject]
                    if (response.containsKey ("error"))
                    {
                        ret.status = RESTfulJSONResult.FAIL
                        ret.message = response.getString ("error")
                        val result = Json.createObjectBuilder
                        for (key <- response.keySet)
                            if (key != "error")
                                result.add (key, response.get (key))
                        ret.setResult (result.build)
                    }
                }
            }
            catch
            {
                case resourceexception: ResourceException =>
                    ret.setResultException (resourceexception, "ResourceException on interaction")
            }
            finally
                try
                    connection.close ()
                catch
                {
                    case resourceexception: ResourceException =>
                        ret.setResultException (resourceexception, "ResourceException on close")
                }
        ret.toString
    }
}

object ListFiles
{
    val LOGGER_NAME: String = ListFiles.getClass.getName
    val _Logger: Logger = Logger.getLogger (LOGGER_NAME)
}
